### PR TITLE
NEW Allow GridFieldEditButton to have configurable HTML classes. Change edit icon.

### PR DESCRIPTION
**Pull request by sachajudd**

### CODE REVIEW

**Comment by tractorcow** [comment]
@sachajudd could you please add a quick before / after screenshot for review? That'd be fab thanks. ;)

**Comment by sachajudd** [comment]
@tractorcow Before: screen shot 2009-13-03 at 1 53 57 pm
After: 
screen shot 2009-13-03 at 1 54 21 pm

File: src/Forms/GridField/GridFieldEditButton.php [hunk 1]
```diff
@@ -95,14 +106,63 @@ public function getColumnContent($gridField, $record, $columnName)
...
     }
 
     /**
+     * Get the extra HTML classes to add for edit buttons
+     *
+     * @return string
+     */
+    public function getExtraClass()
+    {
+        return implode(' ', Convert::raw2att($this->extraClass));
```
Comment (tractorcow) [inline src/Forms/GridField/GridFieldEditButton.php]: The `default_cast` should have already cast these as safely escaped text. The convert is not needed here.

File: src/Forms/GridField/GridFieldEditButton.php [hunk 2]
```diff
@@ -95,14 +106,63 @@ public function getColumnContent($gridField, $record, $columnName)
...
+        return $this;
+    }
+
+    /**
+     * Checks whether the button has the given HTML class
+     *
+     * @param string $class
+     * @return boolean
+     */
+    public function hasExtraClass($class = '')
```
Comment (tractorcow) [inline src/Forms/GridField/GridFieldEditButton.php]: Is this method necessary?

File: src/Forms/GridField/GridFieldEditButton.php [hunk 3]
```diff
@@ -95,14 +106,63 @@ public function getColumnContent($gridField, $record, $columnName)
...
+        return implode(' ', Convert::raw2att($this->extraClass));
+    }
+
+    /**
+     * Add an extra HTML class
+     *
+     * @param string $class
+     * @return $this
+     */
+    public function addExtraClass($class = '')
```
Comment (tractorcow) [inline src/Forms/GridField/GridFieldEditButton.php]: Default arg for addExtraClass should never be used; It's always mandatory right?

File: src/Forms/GridField/GridFieldEditButton.php [hunk 4]
```diff
@@ -95,14 +106,63 @@ public function getColumnContent($gridField, $record, $columnName)
...
+    /**
+     * Remove an HTML class
+     *
+     * @param string $class
+     * @return $this
+     */
+    public function removeExtraClass($class)
+    {
+        $key = array_search($class, $this->extraClass);
+        unset($this->extraClass[$key]);
```
Comment (tractorcow) [inline src/Forms/GridField/GridFieldEditButton.php]: Please check for `$key !== false` before doing unset.

File: src/Forms/GridField/GridFieldEditButton.php (hunk 2 above)
Comment (tractorcow) [reply src/Forms/GridField/GridFieldEditButton.php]: For your tests, I'm happy if you just check getExtraClass has the string (or lack thereof)

File: src/Forms/GridField/GridFieldEditButton.php (hunk 4 above)
Comment (dhensby) [reply src/Forms/GridField/GridFieldEditButton.php]: `unset` doesn't throw errors for undefined vars - so this is safe and going to be slightly faster than adding a condition.

**Comment by tractorcow** [comment]
Rebased, can merge on green.

**Comment by tractorcow** [comment]
Whoops my push failed.

**Comment by sachajudd** [comment]
Thanks, @tractorcow and thanks @robbieaverill for your help 🎉

**Pull request merged**